File: week2/bst.rb
class Node
  attr_reader :key
  attr_accessor :left, :right

  def initialize(key)
    @key = key
    @left = nil
    @right = nil
  end

  def insert(key)
    if self.key > key
      if self.left
        self.left.insert(key)
      else
        self.left = Node.new(key)
      end
    else
      if self.right
        self.right.insert(key)
      else
        self.right = Node.new(key)
      end
    end
  end

  def lookup(key)
    return false unless self
    return self if self.key == key
    if self.key > key
      self.left.loopup(key)
    else
      self.right.loopup(key)
    end
  end

  def delete(key)

  end

  def self_balance
    all_keys = all_nodes.map(&:key).sort
    mid_index = all_keys.length / 2
    self.key = all_keys.slice!(mid_index)
    self.left = nil
    self.right = nil
    all_keys.each do |key|
      node.insert(key)
    end
    node
  end

  def all_nodes
    return unless self
    all = [self]
    all += self.left.all_nodes
    all += self.right.all_nodes
  end

end

File: data_structure/py/bst.py
class Node():
  def __init__(self, val):
    self.val = val
    self.left = None
    self.right = None

  def add(self, node):
    if self.val > node.val:
      if self.left:
        self.left.add(node)
      else:
        self.left = node
    else:
      if self.right:
        self.right.add(node)
      else:
        self.right = node

def TreeToList(node):
  if node is None:
    return []
  left = TreeToList(node.left)
  right = TreeToList(node.right)
  return left + [node.val] + right

n1 = Node(1)
n2 = Node(2)
n3 = Node(3)
n4 = Node(4)
n5 = Node(5)
n6 = Node(6)
n7 = Node(7)
n8 = Node(8)

n4.add(n1)
n4.add(n2)
n4.add(n3)
n4.add(n5)
n4.add(n6)
n4.add(n7)
n4.add(n8)
arr = TreeToList(n4)
print(arr)
